### ISSUE TITLE
GridFieldEditableColumns: callback function on the display fields, first paramter is the incorrect DO

### ISSUE THREAD

**Issue by ElGabbu** [symbiote/silverstripe-gridfieldextensions#198]
DataObjectA (OddsEvent) ... has_many DataObjectB (OddsMatch)

In getCMSField of DataObjectA I have a gridfield showing records for the has_many relationship of DataObjectB. I have added the GridFieldEditableColumns component to this grid field and setup the display fields using the callback function. 

The first parameter of the callback function should always be an instance of DataObjectB ... this is true when displaying the GridField. However during saving of DataObjectA the first param of the callback function is an instance of DataObjectA.

Thanks

**Comment by dhensby** [comment]
Would you mind providing some code to help replicate this issue, please?

**Comment by ElGabbu** [comment]
`$gridFieldConfig->getComponentByType('GridFieldEditableColumns')->setDisplayFields(array(
			'HomeTeamName' => array(
				'title' => 'Home Team',
				'callback' => function($record, $column, $grid) {
					$field = TextField::create($column, 'HomeTeamName');
					if($record->ClassName == 'OddsMatch' && $record->HomeTeam()->exists()){
						return $field->performDisabledTransformation();
					}
					return $field;
				}
			));,`

As you can see above I had to check the ClassName property of the record to avoid an error when calling the HomeTeam() function.

**Comment by dhensby** [comment]
Thanks @ElGabbu, but I need more than that code to understand the scenario in which you experience this problem. In particular I need to understand what classes you have defined, their inheritance and their relationships to one another.

**Comment by ElGabbu** [comment]
@dhensby updated code with classes and their relationships

**Comment by dhensby** [comment]
thanks @ElGabbu but it's fairly clear this isn't representative of the code you're using.

Your `OddsMatch` class should have a `HomeTeam` relation or method and there should be some way to populate the `HomeTeamName` - that plus the fact your `getCMSFields` method is broken (`$fields` is undefined, you don't return `$fields` and it breaks when you try to create a new object because the grid field doesn't exist yet)...

Can you please provide either:
 - Code that is complete and can be used to reproduce the error; OR
 - A test that demonstrates the failure

Thanks

**Comment by ElGabbu** [comment]
@dhensby I prepared some code that demonstrates the issue here: 

https://github.com/ElGabbu/gridfieldextensions-test/tree/master/mysite/code

You should encounter the issue when trying to save a person with a car in the grid field.

**Comment by dhensby** [comment]
Thanks for that - I've been able to reproduce the issue now.

The problem is this:

When saving the record, the record being saved is an instance of `Person` and not of `Car`. Both `GridFieldAddNewInlineButton` and `GridFieldEditableColumns` have `handleSave` methods that are invoked and both create a form instance using the `Person` instance being saved instead of an instance of the related record.

This means the wrong record type is passed to the callback. I've created a PR #200 which fixes the problem.
